### PR TITLE
Decompile unk_02016B94.s

### PR DESCRIPTION
**Pull request by remicalixte**
I managed to decompile the first few functions of this file and then spent way too much time trying to match the next function.

I has a huge switch statement and I couldn't get the right register allocations and order so I'm giving up for now. Are you still interested in my attempt? It still matches a good part of the function but not all of it as I said.

### CODE REVIEW

**Comment by red031000** [comment]
if you are unable to match a function you should ideally use a nonmatching (see an example in text.c) with your best guess

**Comment by remicalixte** [comment]
So I tried to use the ASM_GLOBAL macro and putting the asm function in a file but the build doesn't match this way. Using `asmdiff` I can see that the asm function is injected in the binary but there is some strange padding after it that is shifting everything. I'm trying to understand what `asm_processor.py` is doing

**Comment by red031000** [comment]
if you would like to finish this file, I can merge it in

**Comment by remicalixte** [comment]
I was working on the next function and it's also really hard to match. I tried adding a non-matching as well but asm_processor seems broken when there are several `GLOBAL_ASM` in one file. I'll let this file as is for now and work an easier ones. You can merge the PR now with the asm_processor fixes and the the few decompiled functions if you wish.

I will probably go back on this file one day.

**Comment by remicalixte** [comment]
6000 lines left...

**Comment by remicalixte** [comment]
I finally finished this file. There is a couple of `NONMATCHING` functions.

And then there is `FUN_0201974C`:
- ~2500 lines of assembly
- a stack of 0x314 bytes
- so long that it uses `bl` to branch so the disassembler split it in 3

I think it's some kind of graphics library.

**Review by ProjectRevoTPP — APPROVED** [review]
Let red review this though

File: arm9/src/unk_02016B94.c [hunk 1]
```diff
@@ -0,0 +1,8081 @@
...
+        }
+        else
+        {
+            r3 += param0 + param1 * param2;
+        }
+        break;
+    case 1:
+        if (r5 >= 0)
+        {
+            r3 += (1 << 0xa);
```
Comment (red031000) [inline arm9/src/unk_02016B94.c:2827]: can this be turned into a constant?

**Review by red031000 — CHANGES_REQUESTED** [review]
this

File: arm9/src/unk_02016B94.c (hunk 1 above)
Comment (red031000) [reply arm9/src/unk_02016B94.c:2827]: not constant a number, I'm tired

Comment (remicalixte) [reply arm9/src/unk_02016B94.c]: sure

**Comment by red031000** [comment]
if the conflicts could be resolved I'd be happy to merge this in

**Comment by remicalixte** [comment]
I removed `window.h` during the merge. All the content of the header is actually part of the file as decompiled. I could have renamed the whole file `window.c`/`window.h` but I think it's doing much more than purely window stuff so I don't think it's fitting

**Comment by red031000** [comment]
sorry can you resolve again Kappa

**Review by red031000 — DISMISSED** [review]
approve pending merge conflict resolution

**Comment by remicalixte** [comment]
resolved again

**Review by ProjectRevoTPP — APPROVED** [review]
per red's prior approval

**Pull request merged**